### DISCUSSION TITLE
Hook for when traveling to a search?

### DISCUSSION THREAD

**Discussion by vdegenne** [LazyVim/LazyVim#4443] [Q&A]
I'd like to execute `zt` (to scroll the page until the caret is at the top) every time I accept a search, because sometimes the scroll from search validation will place the caret at the bottom and I always find myself pressing `zt` to reveal the code below.

I have no idea how I can implement that and if there is a general hook for that. Even if it executes every time the caret moves from line to line that would be fine I think, I just want to automate `zt`.

Thanks for the help

**Comment by dpetka2001** [comment]
You could do something like
```lua
return {
  {
    "nvim-telescope/telescope.nvim",
    opts = function(_, opts)
      local action_set = require("telescope.actions.set")

      local picker_config = {}
      for builtin, _ in pairs(builtins) do
        picker_config[builtin] = {
          -- Execute `zt` after selecting an entry from search results
          attach_mappings = function(prompt_bufnr, _)
            action_set.select:enhance({
              post = function()
                vim.cmd(":normal! zt")
              end,
            })

            return true
          end,
        }
      end

      opts.pickers = vim.tbl_deep_extend("force", picker_config, {})

    end,
  }
}
```
Test it to see if it does what you want.

**Reply by vdegenne** [reply]
Thanks @dpetka2001 

I also asked the question on Reddit yesterday (here's the link https://www.reddit.com/r/neovim/comments/1fom55x/hook_for_when_traveling_to_a_search/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button).

Would you consider your solution better? I'll try both methods when I'm at home anyways and give feedback later.

**Reply by dpetka2001** [reply]
That solution utilizes creating an autocmd on `CursorMoved` event. `CursorMoved` is an event that happens quite frequently. It even has a warning on `:h CursorMoved` about not doing expensive stuff there because it might cause slowdowns. My solution uses Telescope's internal states for executing something after you select an entry. `CursorMoved` will happen whenever you move your cursor. Is that what you want or do you want it to happen only after you select a Telescope entry? That's up to you, but keep in mind that the autocmd on `CursorMoved` might cause a slow down, so if you happen to notice anything like that you'd have to remember to disable the autocmd to check difference on performance to investigate if that's the culprit.

**Reply by vdegenne** [reply]
I am not sure yet, all solutions have avantage and disadvantage, I'll have to use yours on some of my slow servers because the solution on reddit seems a lot to process.
I think a more ideal (and cross) solution would be to execute `zt` when the cursor is too low on the screen/page (e.g. > 50%) because after all that's the only reason I'm using this command in the first place. What do you think? Is there an easy way to get the first and last visible line numbers on the page?

**Reply by vdegenne** [reply]
Ok so I ask ChatGPT and the code it gives seems to work correctly,
```lua
local past = vim.loop.now()
local debounce_time = 1000

vim.api.nvim_create_autocmd("CursorMoved", {
  callback = function()
    local now = vim.loop.now()

    if now - past > debounce_time then
      if vim.fn.winline() >= math.floor(vim.api.nvim_win_get_height(0) / 2) then
        vim.cmd("normal! zt")
      end

      past = now
    end
  end,
})
```

**Reply by dpetka2001** [reply]
You can get the window height with `vim.api.nvim_win_get_height(0)`. And you can get the current line relative position from the top of the window with `vim.fn.winline()`. If it's greater than the window's half height you could do `zt`. But this would probably have to be calculated every time you move your cursor, so on `CursorMoved` event, and I don't know if this produces any slowdown as well. I don't really know what the best solution is. You should test them out and see what feels more performant to you. Honestly, I would just do `zt` manually.

But for the Telescope searches, I would personally prefer something that can be done inside Telescope instead of adding an autocmd.

**Reply by dpetka2001** [reply]
Looool, after I commented I saw your new answer which is the same as mine :stuck_out_tongue:

**Reply by vdegenne** [reply]
Yes sorry I was in a hurry 😇 but I always appreciate your help. You can see the edit history on my message, ChatGPT don't always give perfect code the first time and I had to guide it. That's why I like asking help from the community, some of you guys can think optimized code directly.
Btw I don't know if that debouncer is going to make a significant change in perf, I just assume it will.
@dpetka2001 🫡

**Reply by vdegenne** [reply]
@dpetka2001 Hey! Coming back to you. I finally decided to use your solution, it's better to explicitly use `zt` on a specific action.
Also I wanted to bind `zt` to `gd` (go to definition), here's the code:

```lua
map("n", "gd", function()
  vim.lsp.buf.definition()
  vim.defer_fn(function()
    vim.cmd("normal! zt")
  end, 50)
end, opts)
```

However for some reasons the function is not executed when I press `gd` (the function works when I bind it to another keystroke). My guess is that the lsp is rebinding `gd` on load which makes my keymap obsolete.

Ideally now, it would be nice if I could find the right spot in the code responsible for going to the definition OR opening telescope on multiple choices. In other words I need to find the intersection so I can have a model on how to rewrite that code with my own logic.

I wish you could help me with that.

**Reply by dpetka2001** [reply]
Yes, `gd` is a LSP keybinding, so you have to disable it there as well. You could do
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = function()
      local Keys = require("lazyvim.plugins.lsp.keymaps").get()
      vim.list_extend(Keys, {
        -- disable `gd` LSP mapping
        { "gd", false },
        -- make another keymap for `goto_definition`
        { "your_keymap", function() require("telescope.builtin").lsp_definitions({ reuse_win = true }) end, desc = "Goto Definition", has = "definition" },
      })
    end,
  },
}
```
and then in your `keymaps.lua` the following
```lua
vim.keymap.set("n", "gd", "zt", { remap = true })
```
I'm afraid I don't understand what you mean by opening Telescope on multiple choices.

PS: The way to change LSP keymaps is also described in the [docs](https://www.lazyvim.org/plugins/lsp#%EF%B8%8F-customizing-lsp-keymaps). Please have a read through them to understand how to do things.

**Reply by vdegenne** [reply]
Thanks, I'm trying to figure it out with the snippets you gave me.

By the way the first code has some errors:

![image](https://github.com/user-attachments/assets/4d1db380-4e91-466d-a7b0-13f025ca489f)

**Reply by dpetka2001** [reply]
Yeah forgot to add `local builtins = require("telescope.builtin")` for the first error (just put it after `local action_set`). Don't worry about the second one with `undefined field enhance`. And remove the arguments from `attach_mappings = function()`, since they're not actually used.

**Reply by vdegenne** [reply]
Ok it wasn't working and I realized `aerial` was the problem. Not sure why, since `aerial` uses telescope too (symbols picker).

**Reply by dpetka2001** [reply]
Yeah that's because we only change this behavior for the Telescope default builtin pickers. `aerial.nvim` registers its own Telescope picker. I don't know how to modify pickers registered by other plugins. You should ask at Telescope or aerial github repos if there's a way to do that.

**Reply by vdegenne** [reply]
Ok I can live without aerial for now :,-) 
But I'm still not satisfied with this setup, because `zt` is only executed from telescope on `gd` but if there are no multiple choices vim will skip telescope and go to the definition directly, in that case `zt` is ignored too.

I guess I'll have to write my own lsp request and bind `gd` to it.

**Reply by dpetka2001** [reply]
@vdegenne I think I found a solution by experimenting around. For aerial picker you can modify the default keymap like this
```lua
    keys = {
      {
        "<leader>ss",
        function()
          local action_set = require("telescope.actions.set")
          require("telescope").extensions.aerial.aerial({
            attach_mappings = function()
              action_set.select:enhance({
                post = function()
                  vim.cmd(":normal! zt")
                end,
              })

              return true
            end,
          })
        end,
      },
    },
```
It seems to work on my end. Test it out to see if it works for your purposes as well.

PS: this goes in the telescope plugin spec, not the aerial.

**Reply by dpetka2001** [reply]
About when there are not multiple choices, you can add in the pickers' config `jump_type = "never"` to always open the Telescope picker even if there's only one result.

**Reply by dpetka2001** [reply]
See how I had done it when I was still using Telescope [here in my dotfiles](https://github.com/dpetka2001/dotfiles/blob/44cfc4a569b2ef66602140a15121d2962f21e3ec/dot_config/nvim/lua/plugins/telescope.lua#L93-L98).

**Reply by vdegenne** [reply]
@dpetka2001 Nice! I finally have the setup correct now. I use your telescope solution (aerial works yeah) and this is now my `nvim-lspconfig.lua` file:

```lua
-- Custom function to check the behavior of 'vim.lsp.buf.definition()'
local function go_to_definition()
  local params = vim.lsp.util.make_position_params()

  -- Send LSP request for 'textDocument/definition'
  vim.lsp.buf_request(0, "textDocument/definition", params, function(err, result, ctx, _)
    if err then
      vim.notify("Error fetching definition: " .. err.message, vim.log.levels.ERROR)
      return
    end

    if not result or vim.tbl_isempty(result) then
      vim.notify("No definition found", vim.log.levels.INFO)
      return
    end

    if #result == 1 then
      -- Jump directly if only one definition is found
      vim.lsp.util.jump_to_location(result[1], "utf-8")
      vim.cmd("normal! zt")
    else
      -- If multiple definitions are found, open the list
      -- vim.lsp.util.set_qflist(vim.lsp.util.locations_to_items(result))
      -- vim.cmd("copen")
      require("telescope.builtin").lsp_definitions({ reuse_win = true })
    end
  end)
end

return {
  "neovim/nvim-lspconfig",
  opts = function(_, opts)
    opts.inlay_hints = { enable = false }

    local Keys = require("lazyvim.plugins.lsp.keymaps").get()

    vim.list_extend(Keys, {
      {
        "gd",
        go_to_definition,
        desc = "Goto Definition",
        has = "definition",
      },
    })
  end,
  -- opts = {
  --   inlay_hints = { enabled = false },
  -- },
}
```

(ChatGPT gave me the `go_to_definition` function, crazy how it got it right first try)

The lsp api really makes it easy, I learn so much today thanks a lot!

**Reply by vdegenne** [reply]
@dpetka2001 May I ask why you don't use telescope anymore? What do you use as a replacement?

**Reply by dpetka2001** [reply]
FzfLua. I was trying it even when Telescope was the only fuzzy finder in Lazyvim. But when it got available as an Extra i made a full switch to it. People say it's faster in large codebases but i can't attest to that as i'm still new to learning programming (started  at the start of the year) and don't have experience with large codebases. I just like its default minimalistic interface and have also been using fzf in terminal, so i just stuck with it.

**Reply by vdegenne** [reply]
I just tried and it looks very nice, the greyed pathes are cool. Which means now I have to learn how to configure it again fuuu 
I don't see any fzf-lua config in your dotfile repo, I assume you use the default, or you haven't pushed your last changes?

**Reply by dpetka2001** [reply]
Those dotfiles I linked I don't use now so much. I added way too many plugins when I was still learning LazyVim. I don't like it that any more. 

I'm in the process of trying to use mostly stock LazyVim with only the necessary Extras enabled and some utilities of my own (that I still take inspiration around the net obviously and customizing them to my needs). I haven't pushed changes yet to Github of the new LazyVim configuration, because it's still a work in progress and to be honest I don't know if I'm going to finish it. 

The corresponding FzfLua configuration in the new LazyVim configuration, you can see in this [gist](https://gist.github.com/dpetka2001/34a0e4ecc0edb6a9b016bd557da8c0c7).

**Reply by vdegenne** [reply]
Here's how to execute `zt` on item validation for `fzf-lua` plugin:

```lua
      config.defaults.actions.files["enter"] = function(selected, options)
        actions.file_edit(selected, options)
        vim.cmd("normal! zt")
      end
```

**Reply by dpetka2001** [reply]
Nice!! Tnx for sharing this :smile:
